File: backend/app/core/pricing_config.py
"""
PARWA Pricing Configuration — SINGLE SOURCE OF TRUTH
═════════════════════════════════════════════════════════════════════════

This module is the **ONLY** place where pricing data, variant names,
and feature limits are defined. Every other module in the codebase
MUST import from here — never hard-code its own copy.

╔══════════════════════════════════════════════════════════════════╗
║  DO NOT duplicate these constants anywhere else.               ║
║  If you need pricing data → import from pricing_config.        ║
╚══════════════════════════════════════════════════════════════════╝

Pricing model: $1 = 1 ticket. What you pay is how many tickets you get.
All variants have the SAME AI capabilities. The only difference
is the ticket volume restriction.

Correct prices (as of this writing):
  ┌──────────┬──────────────────┬───────────────────┬──────────────┐
  │ Variant  │ Monthly (USD)    │ Annual (USD)      │ Tickets/mo   │
  ├──────────┼──────────────────┼───────────────────┼──────────────┤
  │ Parwa    │ $2,499.00        │ $29,988.00        │ 2,499        │
  │ High     │ $3,999.00        │ $47,988.00        │ 3,999        │
  └──────────┴──────────────────┴───────────────────┴──────────────┘

Mini PARWA was removed on 2026-07-26. Existing Mini subscribers are
auto-upgraded to Parwa via normalize_variant_name().

Annual = 12 × monthly.  NO discounts.  NO free months.

Building Codes:
  BC-002: All money values use Decimal — never float or int.
"""

from decimal import Decimal
from enum import Enum
from typing import Dict, Any


# ══════════════════════════════════════════════════════════════════════════
# VARIANT TYPE ENUM
# ══════════════════════════════════════════════════════════════════════════

class VariantType(str, Enum):
    """
    PARWA subscription variant types.

    Values are the canonical lowercase names: "parwa", "high".

    Mini PARWA was removed on 2026-07-26 — only 2 tiers remain.
    Old names (mini, starter, mini_parwa) are accepted via
    normalize_variant_name() and auto-upgraded to "parwa" (existing
    Mini customers get Parwa features for free — the humane choice).
    """
    PARWA = "parwa"
    HIGH = "high"


# ══════════════════════════════════════════════════════════════════════════
# OLD NAME → CANONICAL NAME MAPPING
# ══════════════════════════════════════════════════════════════════════════

_VARIANT_NAME_ALIASES: Dict[str, str] = {
    # Old name              → Canonical name
    # NOTE: Mini was removed 2026-07-26. Existing Mini subscribers are
    # auto-upgraded to Parwa (more features, same moral commitment).
    "mini_parwa":           "parwa",
    "parwa":                "parwa",
    "parwa_high":           "high",
    # Also accept common variations
    "mini-parwa":           "parwa",
    "parwa-high":           "high",
    "starter":              "parwa",
    "growth":               "parwa",
    "mini":                 "parwa",
    "high":                 "high",
}


# ══════════════════════════════════════════════════════════════════════════
# PRICES — monthly and annual
# ══════════════════════════════════════════════════════════════════════════

VARIANT_PRICES: Dict[VariantType, Decimal] = {
    VariantType.PARWA: Decimal("2499.00"),
    VariantType.HIGH:  Decimal("3999.00"),
}
"""
Monthly prices per variant (USD).

BC-002: All values are Decimal — never float or int.
"""

VARIANT_ANNUAL_PRICES: Dict[VariantType, Decimal] = {
    variant: price * Decimal("12")
    for variant, price in VARIANT_PRICES.items()
}
"""
Annual prices per variant (USD).

Annual = 12 × monthly.  NO discounts.  NO free months.
"""


# ══════════════════════════════════════════════════════════════════════════
# DISPLAY NAMES — for UI/upgrade nudges
# ══════════════════════════════════════════════════════════════════════════

VARIANT_DISPLAY_NAMES: Dict[VariantType, str] = {
    VariantType.PARWA: "PARWA",
    VariantType.HIGH:  "PARWA High",
}


# ══════════════════════════════════════════════════════════════════════════
# VARIANT FEATURE LIMITS
# ══════════════════════════════════════════════════════════════════════════
#
# Pricing model: $1 = 1 ticket.
# All variants have the SAME AI capabilities.
# The only difference is the ticket volume ($1 = 1 ticket).
# Other limits (agents, team members, etc.) also differ by tier.

VARIANT_LIMITS: Dict[VariantType, Dict[str, Any]] = {
    VariantType.PARWA: {
        "monthly_tickets": 2499,
        "ai_agents":       5,
        "team_members":    10,
        "voice_slots":     2,
        "kb_docs":         500,
    },
    VariantType.HIGH: {
        "monthly_tickets": 3999,
        "ai_agents":       8,  # 8 base, +$3/agent overage beyond this
        "team_members":    25,
        "voice_slots":     5,
        "kb_docs":         2000,
    },
}
"""
Feature limits per variant.

Keys per variant:
  monthly_tickets : int   — max support tickets per month ($1 = 1 ticket)
  ai_agents       : int   — max AI agents
  team_members    : int   — max team members
  voice_slots     : int   — max concurrent voice/call slots
  kb_docs         : int   — max knowledge-base documents

Prices are NOT included here — use VARIANT_PRICES or get_variant_price().
This separation prevents accidental use of limits when prices are needed
(and vice-versa).
"""

# Tier ordering for upgrade/downgrade logic
VARIANT_TIER_ORDER: Dict[VariantType, int] = {
    VariantType.PARWA: 1,
    VariantType.HIGH:  2,
}


# ══════════════════════════════════════════════════════════════════════════
# PUBLIC API FUNCTIONS
# ══════════════════════════════════════════════════════════════════════════

def get_variant_price(
    variant: VariantType | str,
    billing_cycle: str = "monthly",
) -> Decimal:
    """
    Get the price for a variant and billing cycle.

    Args:
        variant:       VariantType enum or string name (old or new).
        billing_cycle: "monthly" or "annual".

    Returns:
        Decimal price in USD.

    Raises:
        ValueError: If variant or billing_cycle is invalid.

    Examples:
        >>> get_variant_price("parwa")
        Decimal('2499.00')
        >>> get_variant_price(VariantType.PARWA, "annual")
        Decimal('29988.00')
        >>> get_variant_price("mini")            # legacy → auto-upgraded to parwa
        Decimal('2499.00')
        >>> get_variant_price("parwa_high", "annual")
        Decimal('47988.00')
    """
    # Normalize string to VariantType if needed
    if isinstance(variant, str):
        normalized = normalize_variant_name(variant)
        variant = VariantType(normalized)

    cycle = billing_cycle.lower().strip()
    if cycle == "monthly":
        return VARIANT_PRICES[variant]
    elif cycle == "annual":
        return VARIANT_ANNUAL_PRICES[variant]
    else:
        raise ValueError(
            f"Invalid billing_cycle '{billing_cycle}'. "
            f"Must be 'monthly' or 'annual'."
        )


def normalize_variant_name(name: str) -> str:
    """
    Normalize a variant name to its canonical form.

    Maps old names to new:
      starter     → parwa  (Mini removed; auto-upgraded to Parwa)
      growth      → parwa
      mini_parwa  → parwa  (Mini removed; auto-upgraded to Parwa)
      parwa_high  → high
      mini        → parwa  (Mini removed; auto-upgraded to Parwa)

    Also handles common variations (hyphens, case differences).

    Args:
        name: Variant name string (old or new).

    Returns:
        Canonical lowercase variant name string.

    Raises:
        ValueError: If the name is not recognized.

    Examples:
        >>> normalize_variant_name("starter")
        'parwa'
        >>> normalize_variant_name("parwa_high")
        'high'
        >>> normalize_variant_name("parwa")
        'parwa'
        >>> normalize_variant_name("MINI")
        'parwa'
    """
    if not name or not isinstance(name, str):
        raise ValueError("Variant name must be a non-empty string.")

    key = name.lower().strip()
    canonical = _VARIANT_NAME_ALIASES.get(key)

    if canonical is None:
        raise ValueError(
            f"Unknown variant name '{name}'. "
            f"Valid names: parwa, high. "
            f"Legacy aliases (auto-mapped): starter, growth, mini, mini_parwa, parwa_high."
        )

    return canonical


def get_variant_limits(variant: VariantType | str) -> Dict[str, Any]:
    """
    Get the feature limits for a variant.

    Args:
        variant: VariantType enum or string name (old or new).

    Returns:
        Dict with keys: monthly_tickets, ai_agents, team_members,
        voice_slots, kb_docs.

    Raises:
        ValueError: If the variant is invalid.

    Examples:
        >>> get_variant_limits("parwa")
        {'monthly_tickets': 2499, 'ai_agents': 5, 'team_members': 10, 'voice_slots': 2, 'kb_docs': 500}
        >>> get_variant_limits("parwa_high")["voice_slots"]
        5
    """
    if isinstance(variant, str):
        normalized = normalize_variant_name(variant)
        variant = VariantType(normalized)

    return dict(VARIANT_LIMITS[variant])  # return a copy to prevent mutation


def is_upgrade(old_variant: str, new_variant: str) -> bool:
    """
    Check whether moving from old_variant to new_variant is an upgrade.

    Args:
        old_variant: Current variant name (old or new format).
        new_variant: Target variant name (old or new format).

    Returns:
        True if new_variant is a higher tier than old_variant.

    Raises:
        ValueError: If either variant name is invalid.
    """
    old_norm = VariantType(normalize_variant_name(old_variant))
    new_norm = VariantType(normalize_variant_name(new_variant))
    return VARIANT_TIER_ORDER[new_norm] > VARIANT_TIER_ORDER[old_norm]


def get_all_variant_info(variant: VariantType | str) -> Dict[str, Any]:
    """
    Get comprehensive info for a variant: price, limits, and display name.

    Convenience function that combines get_variant_price and get_variant_limits.

    Args:
        variant: VariantType enum or string name (old or new).

    Returns:
        Dict with: variant, display_name, monthly_price, annual_price,
        monthly_tickets, ai_agents, team_members, voice_slots, kb_docs.
    """
    if isinstance(variant, str):
        normalized = normalize_variant_name(variant)
        variant = VariantType(normalized)

    return {
        "variant":        variant.value,
        "display_name":   VARIANT_DISPLAY_NAMES[variant],
        "monthly_price":  VARIANT_PRICES[variant],
        "annual_price":   VARIANT_ANNUAL_PRICES[variant],
        **VARIANT_LIMITS[variant],
    }

File: src/lib/pricing-config.ts
/**
 * PARWA Pricing Configuration — Frontend Single Source of Truth
 * ═══════════════════════════════════════════════════════════════════
 *
 * This module is the ONLY place where pricing data and variant info
 * should be defined on the frontend. Every other component MUST
 * import from here — never hard-code its own copy.
 *
 * Matches backend: /backend/app/core/pricing_config.py
 *
 * ╔══════════════════════════════════════════════════════════════════╗
 * ║  DO NOT duplicate these constants anywhere else.               ║
 * ║  If you need pricing data → import from pricing-config.        ║
 * ╚══════════════════════════════════════════════════════════════════╝
 *
 * Pricing model: $1 = 1 ticket. What you pay is how many tickets you get.
 *   Parwa:  $2,499/mo = 2,499 tickets/mo  | $29,988/yr
 *   High:   $3,999/mo = 3,999 tickets/mo  | $47,988/yr
 *
 * All variants have the SAME AI capabilities. The only difference
 * is the ticket volume restriction.
 *
 * Mini PARWA was removed on 2026-07-26. Existing Mini subscribers are
 * auto-upgraded to Parwa via normalizeTier().
 *
 * Annual = 12 × monthly. NO discounts. NO free months.
 *
 * Building Codes:
 *   BC-002: All money values use number (cents not needed on frontend,
 *           but never use float arithmetic for billing calculations)
 */

// ── Variant Types ──────────────────────────────────────────────────

export type VariantTier = 'parwa' | 'high';

// ── Prices (Monthly USD) ───────────────────────────────────────────

export const VARIANT_PRICES: Record<VariantTier, number> = {
  parwa: 2499,
  high: 3999,
};

export const VARIANT_ANNUAL_PRICES: Record<VariantTier, number> = {
  parwa: 2499 * 12,   // 29,988
  high: 3999 * 12,    // 47,988
};

// ── Display Names ──────────────────────────────────────────────────

export const VARIANT_DISPLAY_NAMES: Record<VariantTier, string> = {
  parwa: 'PARWA',
  high: 'PARWA High',
};

export const VARIANT_TAGLINES: Record<VariantTier, string> = {
  parwa: 'The Junior Agent',
  high: 'The Senior Agent',
};

// ── Variant Limits — $1 = 1 ticket ────────────────────────────────

export interface VariantLimits {
  monthlyTickets: number;
  aiAgents: number;
  teamMembers: number;
  voiceSlots: number;
  kbDocs: number;
}

/**
 * Pricing model: $1 = 1 ticket.
 * What you pay = how many tickets you get.
 * All variants have the SAME AI capabilities — only ticket volume differs.
 */
export const VARIANT_LIMITS: Record<VariantTier, VariantLimits> = {
  parwa: {
    monthlyTickets: 2499,
    aiAgents: 5,
    teamMembers: 10,
    voiceSlots: 2,
    kbDocs: 500,
  },
  high: {
    monthlyTickets: 3999,
    aiAgents: 8, // 8 base, +$3/agent overage
    teamMembers: 25,
    voiceSlots: 5,
    kbDocs: 2000,
  },
};

// ── AI Pipeline Info ──────────────────────────────────────────────
// All variants have the SAME capabilities. Only ticket volume differs.

export const VARIANT_AI_INFO: Record<VariantTier, {
  pipelineSteps: number;
  aiResolution: number;
  techniques: string;
  concurrentCalls: number;
}> = {
  parwa: {
    pipelineSteps: 9,
    aiResolution: 0.88,
    techniques: 'All 14 techniques (Tier 1+2+3)',
    concurrentCalls: 5,
  },
  high: {
    pipelineSteps: 9,
    aiResolution: 0.88,
    techniques: 'All 14 techniques (Tier 1+2+3)',
    concurrentCalls: 5,
  },
};

// ── Tier Ordering ──────────────────────────────────────────────────

export const VARIANT_TIER_ORDER: Record<VariantTier, number> = {
  parwa: 1,
  high: 2,
};

/**
 * Normalize any variant string (including legacy names) to a canonical VariantTier.
 * Mini PARWA was removed 2026-07-26 — legacy 'mini'/'starter'/'mini_parwa' are
 * auto-upgraded to 'parwa' (existing Mini customers get Parwa features for free).
 */
const LEGACY_TIER_MAP: Record<string, VariantTier> = {
  parwa: 'parwa',
  high: 'high',
  // Legacy aliases → auto-upgraded to parwa
  mini: 'parwa',
  starter: 'parwa',
  growth: 'parwa',
  pro: 'parwa',
  mini_parwa: 'parwa',
  parwa_high: 'high',
};

export function normalizeTier(variant: string): VariantTier {
  const lower = variant.toLowerCase().trim();
  return LEGACY_TIER_MAP[lower] || 'parwa';
}

export function isUpgrade(oldTier: VariantTier, newTier: VariantTier): boolean {
  return VARIANT_TIER_ORDER[newTier] > VARIANT_TIER_ORDER[oldTier];
}

// ── Add-Ons ────────────────────────────────────────────────────────

export interface AddOn {
  key: 'voice' | 'customApi';
  name: string;
  description: string;
  price: number;
  /** Which variants already include this feature (no extra charge) */
  includedIn: VariantTier[];
}

export const ADD_ONS: AddOn[] = [
  {
    key: 'voice',
    name: 'Voice Channel',
    description: 'AI-powered inbound & outbound voice calls with real-time transcription.',
    price: 199,
    includedIn: ['parwa', 'high'],
  },
  {
    key: 'customApi',
    name: 'Custom API Connector',
    description: 'Connect any REST API with custom auth and schema mapping.',
    price: 49,
    includedIn: ['parwa', 'high'],
  },
];

// ── Overage Pricing ────────────────────────────────────────────────
// $1 per ticket over limit (same as base rate)

export const OVERAGE_PRICE_PER_TICKET = 1.00; // $1 per ticket over limit

// ── Human Cost Benchmarks (for savings calculation, matches ROI Calculator) ──

export const AGENT_COST_MONTHLY = 4500;
export const TICKETS_PER_AGENT = 400;

// ── Helper: Calculate Overage ──────────────────────────────────────

export function calculateOverage(
  ticketsUsed: number,
  activeVariants: VariantTier[]
): { totalOverageTickets: number; overageCost: number; totalTicketLimit: number } {
  const totalTicketLimit = activeVariants.reduce(
    (sum, tier) => sum + VARIANT_LIMITS[tier].monthlyTickets,
    0
  );
  const totalOverageTickets = Math.max(0, ticketsUsed - totalTicketLimit);
  const overageCost = totalOverageTickets * OVERAGE_PRICE_PER_TICKET;
  return { totalOverageTickets, overageCost, totalTicketLimit };
}

// ── Helper: Calculate Monthly Total ────────────────────────────────

export interface CostBreakdownResult {
  baseSubscription: number;
  addOns: { voice: number; customApi: number };
  overageCost: number;
  totalMonthly: number;
  totalAnnual: number;
  savingsVsHumans: number;
  savingsPercent: number;
  agentsReplaced: number;
  totalTicketLimit: number;
}

export function calculateCostBreakdown(
  activeVariants: VariantTier[],
  addOns: { voice: boolean; customApi: boolean },
  ticketsUsed: number
): CostBreakdownResult {
  // Base subscription
  const baseSubscription = activeVariants.reduce(
    (sum, tier) => sum + VARIANT_PRICES[tier],
    0
  );

  // Add-ons (only charge if not included in any active variant)
  const voiceIncluded = activeVariants.some(t => ADD_ONS[0].includedIn.includes(t));
  const customApiIncluded = activeVariants.some(t => ADD_ONS[1].includedIn.includes(t));
  const voiceCost = addOns.voice && !voiceIncluded ? ADD_ONS[0].price : 0;
  const customApiCost = addOns.customApi && !customApiIncluded ? ADD_ONS[1].price : 0;

  // Overage
  const { overageCost, totalTicketLimit } = calculateOverage(ticketsUsed, activeVariants);

  // Total
  const totalMonthly = baseSubscription + voiceCost + customApiCost + overageCost;
  const totalAnnual = totalMonthly * 12;

  // Savings vs humans
  const agentsReplaced = Math.max(1, Math.round(totalTicketLimit / TICKETS_PER_AGENT));
  const humanCost = agentsReplaced * AGENT_COST_MONTHLY;
  const savingsVsHumans = Math.max(0, humanCost - totalMonthly);
  const savingsPercent = humanCost > 0 ? Math.round((savingsVsHumans / humanCost) * 100) : 0;

  return {
    baseSubscription,
    addOns: { voice: voiceCost, customApi: customApiCost },
    overageCost,
    totalMonthly,
    totalAnnual,
    savingsVsHumans,
    savingsPercent,
    agentsReplaced,
    totalTicketLimit,
  };
}

// ── Helper: "What If" Upgrade Preview ──────────────────────────────

export interface WhatIfPreview {
  currentTotal: number;
  newTotal: number;
  difference: number;
  newTicketLimit: number;
  ticketIncrease: number;
  savingsVsHumans: number;
}

export function calculateWhatIfUpgrade(
  currentVariants: VariantTier[],
  addOns: { voice: boolean; customApi: boolean },
  ticketsUsed: number,
  targetTier: VariantTier
): WhatIfPreview {
  const current = calculateCostBreakdown(currentVariants, addOns, ticketsUsed);

  // Add the target tier if not already present
  const newVariants = currentVariants.includes(targetTier)
    ? currentVariants
    : [...currentVariants, targetTier];
  const upgraded = calculateCostBreakdown(newVariants, addOns, ticketsUsed);

  return {
    currentTotal: current.totalMonthly,
    newTotal: upgraded.totalMonthly,
    difference: upgraded.totalMonthly - current.totalMonthly,
    newTicketLimit: upgraded.totalTicketLimit,
    ticketIncrease: upgraded.totalTicketLimit - current.totalTicketLimit,
    savingsVsHumans: upgraded.savingsVsHumans,
  };
}